### ISSUE TITLE
Retina+jpeg option?

### ISSUE THREAD

**Issue by nspies-celsiustx** [ipython/matplotlib-inline#5]
I often work with complex matplotlib figures with millions of points, which I've been displaying in my notebooks using the ipython magic `%config InlineBackend.figure_format = 'retina'`. Unfortunately, these inline images are quite large in terms of file size.

I'm looking for general suggestions on how to reduce the size of my ipynb files, which can grow into the 10s or even hundreds of MB in size and become a bit unwieldy.

One thought I had was to switch to generating jpegs, but while it looks like I can meaningfully reduce the file size by switching the backend to jpeg and increasing the dpi to 2x:

```python
from matplotlib import rcParams
%config InlineBackend.figure_format = 'jpeg'
rcParams['figure.dpi'] = 200
```

From this repo, it looks like I can dial in jpeg compression values, which should help reduce file size. However, I can't figure out how to size the jpegs appropriately for a retina display (they're twice as large as they should be, and slightly blurry). Is there a way to display jpegs in jupyter notebooks at similar resolution and size to the defaults provided by `'retina'` mode?

See also:
[Matplotlib issue](https://github.com/matplotlib/matplotlib/issues/20697)
[Stackoverflow question](https://stackoverflow.com/questions/68429854/reduce-file-size-of-matplotlib-images-in-jupyter-notebooks)

**Comment by minrk** [comment]
This isn't an expressed option. It certainly could be, but we never added it because jpg compression isn't a great fit for the sharp lines and text on most matplotlib charts.

You can make 2x figures by adding:

```python
def retina_jpeg(figure):
    """Renderer for 2x jpeg images"""
    jpeg_bytes = print_figure(figure, fmt="jpg")
    w, h = _jpegxy(jpeg_bytes)
    return (jpeg_bytes, {"width": w // 2, "height": h // 2})


def enable_retina_jpeg():
    """Tell IPython to use the 2x jpeg renderer"""
    display_formatter = get_ipython().display_formatter
    # disable any existing Figure formatters
    [f.pop(Figure, None) for f in display_formatter.formatters.values()]
    # register 2x jpg
    jpg_formatter = display_formatter.formatters["image/jpeg"]
    jpg_formatter.for_type(Figure, retina_jpeg)
```

Since `_jpegxy` is a private method, you can [copy the implementation](https://github.com/ipython/ipython/blob/c223896f57fff7ca632eb23deb71282eb51f415c/IPython/core/display.py#L766-L783) to be extra safe, but I don't think it's going anywhere.

Here's [a notebook][] showing a complete example.

[a notebook]: https://gist.github.com/fcf1b52199f9cc2e4fd1687ba530190e

But I think the 'right' way to do it here is to separate the configuration from a single 'retina' vs 'png' *or* 'jpg' option to a separate 'scale' option. Where e.g. a figure dpi and screen dpi are separate values, and we return the metadata with the appropriate scale.

**Comment by mgeier** [comment]
@nspies-celsiustx Did you try using SVG with rasterization?

Theoretically, this would give you sharp edges for lines and text etc., but still a reduced size.

See https://matplotlib.org/stable/gallery/misc/rasterization_demo.html

**Comment by nspies-celsiustx** [comment]
> @nspies-celsiustx Did you try using SVG with rasterization?
> 
> Theoretically, this would give you sharp edges for lines and text etc., but still a reduced size.
> 
> See https://matplotlib.org/stable/gallery/misc/rasterization_demo.html

Thanks for the suggestion @mgeier - this sounds like a good option for one-offs but I'm looking for a solution that works for every graphic generated without additional boilerplate.

**Comment by flying-sheep** [comment]
Yeah, let’s separate this out. IPython has [_retina_shape](https://github.com/ipython/ipython/blob/50fa9b942878f6c53d179b1b5151054255a8097f/IPython/core/display.py#L1040), so
- that implementation should go to a spot where both matplotlib_inline and IPython can use it
- we should add WebP support to it, it’s not 1995 (i.e. JPEG is *not* the best format for bandwidth-friendly images on the web anymore)

Where would be a good spot for that code to live?

**Comment by flying-sheep** [comment]
For the time being I’m enabling the existing WebP support in IPython: https://github.com/ipython/ipython/pull/15276
